text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,880 @@
+- Feature Name: `msrv-resolver`
+- Start Date: 2023-11-14
+- Pre-RFC: [internals](https://internals.rust-lang.org/t/pre-rfc-msrv-aware-resolver/19871)
+- RFC PR: [rust-lang/rfcs#3537](https://github.com/rust-lang/rfcs/pull/3537)
+- Rust Issue: [rust-lang/cargo#9930](https://github.com/rust-lang/cargo/issues/9930)
+
+# Summary
+[summary]: #summary
+
+Provide a happy path for developers needing to work with older versions of Rust by
+- Preferring MSRV (minimum-supported-rust-version) compatible dependencies when Cargo resolves dependencies
```

Comment:
@Darksonn 

> The fact that resolver.precedence=rustc is not the default option is very surprising to me, and it's something I missed when I first looked at this RFC.

*(moved from [here](https://github.com/rust-lang/rfcs/pull/3537#pullrequestreview-1833961646) to avoid interleaved conversations)*

Reply:
Could you help me understand where this confusion came from?  Through the curse of knowledge, I'm having a hard time understanding how that happened and would like to see how I could improve the RFC.

Reply:
When I specify an older MSRV in my Cargo.toml, I generally have two configurations that I test in CI:

1. Latest Rust with latest dependencies.
2. MSRV rust with MSRV-compatible dependencies.

The first configuration is the most important one, since that is what most of the users will be using. So generally, I would have one CI run with MSRV-Rust, and run all other CI jobs with latest Rust. I usually don't check in a Cargo.lock, but if I did, then I would check in the first configuration, as it is the most important one.

Therefore, it seems weird to me that the tool would pick MSRV-compatible dependencies when I am using the latest Rust, as that is a signal that I'm currently using the latest-dependencies configuration.

As for how you could have changed the RFC so that I didn't miss this ... I admit to having only skimmed it the first time. I simply saw the RFC, got excited because this is something I've wanted for a long time, and made the assumption that the behavior matched what I would have written myself. If you had included some examples of the new behavior near the beginning, I would probably have noticed there.

Tokio currently relies on the fact that "latest dependencies" == "MSRV-compatible dependencies" and removes dependencies that break this property. This feature would let us remove that requirement, so I am very much looking forward to it.